### PR TITLE
Perf 671: create a get_one_or_create method

### PR DESCRIPTION
**Pull request by olel-gyg**
- Added a `get_one_or_create` method.

### CODE REVIEW

File: tests/ad_templates/models/test_get_one_or_create.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+from tests.ad_templates.factories import EtaAdTemplateFactory, AdLabelFactory, AdPartFactory
+from tests.base import BaseTestCase
+from tms.ad_templates.models import EtaAdTemplate
+
+
+class TestGetOneOrCreate(BaseTestCase):
+
+    def test_should_save_model_obj_from_kwargs(self):
+        headline1=AdPartFactory(template_text="headline_1")
```
Comment (olga-andreeva-gyg) [inline tests/ad_templates/models/test_get_one_or_create.py]: Spaces missing.

File: tests/ad_templates/models/test_get_one_or_create.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        headline2=AdPartFactory(template_text="headline_2")
+        description=AdPartFactory(template_text="description")
+        path1=AdPartFactory(template_text="path_1")
+        path2=AdPartFactory(template_text="path_2")
+        ad_label=AdLabelFactory()
+
+        filters = {
```
Comment (olga-andreeva-gyg) [inline tests/ad_templates/models/test_get_one_or_create.py]: Can be passed directly to `EtaAdTemplate`. Used one there anyway.

File: tests/ad_templates/models/test_get_one_or_create.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        found_eta_ad_template = EtaAdTemplate(
+            headline1=headline1,
+            headline2=headline2,
+            description=description,
+            path1=path1,
+            path2=path2,
+            ad_label=ad_label
+        ).get_one_or_create()
+
+        self.assertEqual(eta_ad_template, found_eta_ad_template)
```
Comment (olga-andreeva-gyg) [inline tests/ad_templates/models/test_get_one_or_create.py]: Would be nice to test the number of records in the table.

File: tms/ad_templates/resources/spreadsheet.py [hunk 4]
```diff
@@ -63,8 +63,15 @@ def get(self):
                     'errors': err.messages,
                 })
             else:
-                db_id = serialized_ad_part.data.get_id()
-                ad_part_id_to_db_id_map[ad_part['id']] = db_id
+                data = serialized_ad_part.data
```
Comment (olga-andreeva-gyg) [inline tms/ad_templates/resources/spreadsheet.py]: Should be consistent with lines 91-100, i.e. maybe name the variable `serialized_ad_part_data` and use it on the next line.

File: tms/ad_templates/resources/spreadsheet.py [hunk 5]
```diff
@@ -63,8 +63,15 @@ def get(self):
...
+                ad_part_obj = serialized_ad_part.data.get_one_or_create(**{
```
Comment (olga-andreeva-gyg) [inline tms/ad_templates/resources/spreadsheet.py]: Can pass arguments directly, without the dict. Also on line 92.

File: tms/ad_templates/resources/spreadsheet.py [hunk 6]
```diff
@@ -81,10 +88,20 @@ def get(self):
...
+                    'path1_id': serialized_template_data.path1_id,
+                    'path2_id': serialized_template_data.path2_id,
+                    'ad_label_id': serialized_template_data.ad_label_id,
+                    'language_id': serialized_template_data.language_id
+                })
                 num_of_valid_eta_ad_templates += 1
-
-        response['eta_ad_templates']['result'] = (f'{num_of_valid_eta_ad_templates} out of {len(eta_ad_templates)} '
-                                          f'processed successfully.')
+        #
```
Comment (olga-andreeva-gyg) [inline tms/ad_templates/resources/spreadsheet.py]: Should be removed.

File: tms/database.py [hunk 7]
```diff
@@ -22,28 +23,25 @@ def scoped_session():
...
         Method to generate a dataframe from a Model
         """
         return pd.read_sql_table(self.__tablename__, db.session.get_bind())
 
+    def get_one_or_create(self, **kwargs):
+        try:
+            instance = self.query.filter_by(**kwargs).one()
+        except orm.exc.NoResultFound as no_result_found_error:
+            self.save()
+            instance = self #? Does this make sense? Definitely a hack!
```
Comment (olga-andreeva-gyg) [inline tms/database.py]: Looks fine to me :)

File: tms/ad_templates/schemas.py [hunk 8]
```diff
@@ -100,7 +100,7 @@ def validate_ad_parts(self, data):
 
     @post_load
     def make_eta_ad_template(self, data):
-        ad_label_id = AdLabel(name=data['ad_label']).get_id()
+        ad_label_id = AdLabel(name=data['ad_label']).get_one_or_create(**{'name': data['ad_label']}).id
```
Comment (olga-andreeva-gyg) [inline tms/ad_templates/schemas.py]: Dict not needed.

File: tests/ad_templates/api/test_generate_gdoc.py [hunk 9]
```diff
@@ -54,7 +54,7 @@ def test_get_one_template(self):
                     'errors': [],
                 }
             })
-
+            #
```
Comment (olga-andreeva-gyg) [inline tests/ad_templates/api/test_generate_gdoc.py]: Should be removed.

File: manage.py [hunk 10]
```diff
@@ -17,10 +17,16 @@
 @manager.command
 def db_create():
     logger.info('initialising database to app and creating missing tables')
-    # Create missing databases and/or tables in the db platform being used
     db.create_all(app=app)
 
 
+@manager.command
```
Comment (andrebco) [inline manage.py]: Isnt dangerous to have a command like that without any check of environment or a confirmation step?

File: tests/ad_templates/models/test_get_one_or_create.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        ad_label = AdLabelFactory()
+
+        eta_ad_template = EtaAdTemplateFactory(
+            headline1=headline1,
+            headline2=headline2,
+            description=description,
+            path1=path1,
+            path2=path2,
+            ad_label=ad_label
+        )
```
Comment (andrebco) [inline tests/ad_templates/models/test_get_one_or_create.py]: Couldnt you use `EtaAdTemplateFactory` right away?

File: manage.py (hunk 10 above)
Comment (olel-gyg) [reply manage.py]: Yes, it's quite dangerous. I'll add a warning message ;-)

**Comment by olel-gyg** [comment]
@andrebco @olga-andreeva-gyg PRA

File: tms/ad_templates/resources/spreadsheet.py [hunk 12]
```diff
@@ -63,22 +63,20 @@ def get(self):
...
             else:
-                data = serialized_ad_part.data
-                ad_part_obj = serialized_ad_part.data.get_one_or_create(**{
-                    'template_text': data.template_text,
-                    'template_type': data.template_type,
-                    'landing_page_type': data.landing_page_type,
-                    'language_id': data.language_id,
-                    'is_fallback': data.is_fallback,
-                })
+                serialised_ad_part_data = serialized_ad_part.data
```
Comment (olga-andreeva-gyg) [inline tms/ad_templates/resources/spreadsheet.py]: serialised or serialized?

File: tms/database.py [hunk 13]
```diff
@@ -21,33 +22,27 @@ def scoped_session():
 
 
 class BaseModel:
-
-    def exists(self):
-        raise NotImplementedError()
+    created = db.Column(db.DateTime, default=datetime.utcnow())
```
Comment (olga-andreeva-gyg) [inline tms/database.py]: Now Flask-Migrate wants to generate a new migration...

File: manage.py (hunk 10 above)
Comment (andrebco) [reply manage.py]: maybe also a `confirm = input('Are you sure?')`

**Pull request merged**